### PR TITLE
Add raw irept dump of lhs to trace

### PR DESCRIPTION
**Pull request by NathanJPhillips**
Add a command-line switch to optionally add a raw `irept` dump of the lhs of assignments to the trace output.
This can be used for diagnostic purposes, for an intelligent UI to display a better interpretation of traces or for post-processing of traces.

### CODE REVIEW

File: src/goto-programs/json_goto_trace.cpp [hunk 1]
```diff
@@ -29,7 +30,8 @@ Author: Daniel Kroening
 void convert(
   const namespacet &ns,
   const goto_tracet &goto_trace,
-  jsont &dest)
+  jsont &dest,
+  bool useExtendedFormat)
```
Comment (smowton) [inline src/goto-programs/json_goto_trace.cpp]: Underscores not camel-case

File: src/goto-programs/json_goto_trace.h [hunk 2]
```diff
@@ -21,6 +21,7 @@ Date: November 2005
 void convert(
   const namespacet &,
   const goto_tracet &,
-  jsont &);
+  jsont &,
+  bool useExtendedFormat = false);
```
Comment (smowton) [inline src/goto-programs/json_goto_trace.h]: same

**Review by smowton — COMMENTED** [review]
Looks good, don't use camel-case in cbmc, and please add a basic regression test or two.

**Review by allredj — COMMENTED** [review]
I don't have much to say about the implementation, but I have two general remarks:
- Could you please add a rationale for that change (at least in the commit or the PR description)? Is it for debug purposes or is this something that we want to use in the product?
- Could we extend the "--trace" option instead of adding an entirely new option (there's already a lot of them)? I can't think of a case where we would use the `--extended-trace` option without the `--trace` option.

**Comment by NathanJPhillips** [comment]
@allredj - apparently we can't keep the current `--trace` and also support `--trace=extended`, which would be nice, I agree.
I've added some possible rationales in the description, though the use-cases this is being written for are actually company-specific so not relevant to the open-source product.

**Comment by allredj** [comment]
@NathanJPhillips  Thanks for the rationale. Pity we can't improve the option. The code looks fine to me but I don't know this part well enough to judge of its sanity, so I'll leave that unapproved.

**Comment by tautschnig** [comment]
Please synchronise with #1565.

**Comment by tautschnig** [comment]
> I've added some possible rationales in the description, though the use-cases this is being written for are actually company-specific so not relevant to the open-source product.

Could the information please be included in the commit messages? https://chris.beams.io/posts/git-commit/ - thanks.

**Comment by NathanJPhillips** [comment]
> Please synchronise with #1565.

Does this mean this PR is effectively blocked for now?

**Comment by tautschnig** [comment]
> Does this mean this PR is effectively blocked for now?

Why would that be the case? You're free to engage on #1565 in any way you feel is appropriate, and if @polgreen happens to be unavailable then you can just pick up her commits. (I think that sort of engagement should actually have happened well before posting this PR, TBH.)

**Comment by polgreen** [comment]
I can make some time to resurrect that PR. The thing i probably don't have time to do is integrating the trace output changes into jbmc (smowton has an outline of how to do it in the PR).  I can address the other comments.

**Comment by smowton** [comment]
@tautschnig as far as I can tell there's no dependency between the two PRs, except to use a matching naming convention for the command-line options. @polgreen has e.g. `--trace-hex`, which seems reasonable -- @NathanJPhillips should perhaps therefore use `--trace-json-extended` for consistency. Otherwise I think they are independent functional additions and it should not matter which goes in first.

**Comment by tautschnig** [comment]
> @tautschnig as far as I can tell there's no dependency between the two PRs, except to use a matching naming convention for the command-line options. 

I had't claimed there was any dependency. But, as you say, matching naming conventions would be a good start. The second step would be synchronising those additional options across cbmc/jbmc front-ends (each of the two PRs does only one of them).

**Comment by NathanJPhillips** [comment]
This PR now covers CBMC and JBMC and has the facility to be extended with more options.

File: src/cbmc/all_properties.cpp [hunk 3]
```diff
@@ -219,7 +219,11 @@ void bmc_all_propertiest::report(const cover_goalst &cover_goals)
         if(g.second.status==goalt::statust::FAILURE)
         {
           jsont &json_trace=result["trace"];
-          convert(bmc.ns, g.second.goto_trace, json_trace);
+          convert(
+            bmc.ns,
+            g.second.goto_trace,
+            json_trace,
+            bmc.trace_options());
```
Comment (tautschnig) [inline src/cbmc/all_properties.cpp]: Why is that only enabled for --all-properties?

File: src/cbmc/bmc.h [hunk 4]
```diff
@@ -98,6 +100,13 @@ class bmct:public safety_checkert
   virtual void show_vcc_plain(std::ostream &out);
   virtual void show_vcc_json(std::ostream &out);
 
+  trace_optionst trace_options()
+  {
+    return trace_optionst {
+      options.get_bool_option("extended-trace"),
+    };
+  }
```
Comment (tautschnig) [inline src/cbmc/bmc.h]: Could options parsing please be moved to a single, common location? I'd just make this a member of `trace_optionst`.

File: src/goto-programs/json_goto_trace.h [hunk 5]
```diff
@@ -21,6 +21,7 @@ Date: November 2005
 void convert(
   const namespacet &,
   const goto_tracet &,
-  jsont &);
+  jsont &,
+  trace_optionst trace_options = trace_optionst::default_options);
```
Comment (tautschnig) [inline src/goto-programs/json_goto_trace.h:25]: I'd suggest not to use default arguments here. That would make any call sites that haven't be adjusted fail compilation.

**Comment by NathanJPhillips** [comment]
Basic regression tests added as per [request](https://github.com/diffblue/cbmc/pull/1694#pullrequestreview-86398427) by @smowton

**Comment by smowton** [comment]
@tautschnig OK for this to go in?

File: src/goto-programs/json_goto_trace.cpp [hunk 6]
```diff
@@ -94,6 +96,39 @@ void convert(
...
+            explicit comment_base_name_visitort(const namespacet &ns) : ns(ns)
+            { }
+
+            void operator()(exprt &expr) override
+            {
+              if(expr.id() == ID_symbol)
+              {
+                const symbolt &symbol = ns.lookup(expr.get(ID_identifier));
+                // Don't break sharing unless need to write to it
+                const irept::named_subt &comments = expr.get_comments();
```
Comment (tautschnig) [inline src/goto-programs/json_goto_trace.cpp]: Sharing will be broken in multiple ways: `expr` is non-const, thus a non-const member will be called; and sharing has been broken at the prior level already by invoking `visit()`.

File: src/cbmc/bmc.h [hunk 7]
```diff
@@ -98,6 +100,13 @@ class bmct:public safety_checkert
   virtual void show_vcc_plain(std::ostream &out);
   virtual void show_vcc_json(std::ostream &out);
 
+  trace_optionst trace_options()
+  {
+    return trace_optionst {
+      options.get_bool_option("trace-json-extended"),
+    };
+  }
```
Comment (tautschnig) [inline src/cbmc/bmc.h]: Copying from a previous comment: Could options parsing please be moved to a single, common location? I'd just make this a member of trace_optionst.

**Review by smowton — APPROVED** [review]

File: src/cbmc/all_properties.cpp [hunk 8]
```diff
@@ -219,7 +219,7 @@ void bmc_all_propertiest::report(const cover_goalst &cover_goals)
         if(g.second.status==goalt::statust::FAILURE)
         {
           jsont &json_trace=result["trace"];
-          convert(bmc.ns, g.second.goto_trace, json_trace);
+          convert(bmc.ns, g.second.goto_trace, json_trace, bmc.trace_options());
```
Comment (tautschnig) [inline src/cbmc/all_properties.cpp:222]: My question still stands: why is this only being done for that particular invocation of `convert`, and not for the one in `bmc.cpp`?

**Comment by smowton** [comment]
@tautschnig this is now done in `bmc.cpp` too for symmetry -- happy for this to go in?

**Comment by tautschnig** [comment]
@NathanJPhillips @smowton I'd prefer not to have default arguments here (i.e., in `convert`) to make sure people intentionally choose the default settings. If you strongly disagree then go ahead and merge as-is, else fix first and then merge.

**Comment by smowton** [comment]
I disagree on that one, I think unadjusted callsites will gain a sane default, and someone using it will additionally be given guidance about a sane default choice without having to search. Therefore I'll merge as-is.

**Pull request merged**